stackrox/stackrox
# ROX-20184: deduper state pagination

## Description

If the `DeduperState` has more than `ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE` (default 200k) entries Central sends the `DeduperState` in chunks of a maximum of `ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE` entries per chunk. The `DeduperState` has two new fields:
* `Total`: Indicates the total number of chunks that will be sent.
* `Current`: Indicates  from 1 to `Total` which chunk was sent.

Sensor expects the chunks sent in order. If Sensor does not receive the chunks in order it restarts the connection with Central and defaults to not doing reconciliation in Sensor.

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

  * Deploy ACS in a cluster
  * Set ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE=10: `ks set env deploy/central ROX_MAX_DEDUPER_ENTRIES_PER_MESSAGE=10`
  * On reconnect you should see the deduper chunks are being sent in central and sensor's logs:
```
# Central logs: after reconnect
sensor/service/connection: 2023/10/20 09:57:42.495441 connection_impl.go:605: Info: Sensor (6cfe4d44-48d0-498a-a0a3-a0c39c832d6a) can do client reconciliation: sending deduper state
sensor/service/connection: 2023/10/20 09:57:42.495760 connection_impl.go:704: Info: Sending 10 hashes (Size=603), current chunk: 1, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496014 connection_impl.go:704: Info: Sending 10 hashes (Size=587), current chunk: 2, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496197 connection_impl.go:704: Info: Sending 10 hashes (Size=607), current chunk: 3, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496398 connection_impl.go:704: Info: Sending 10 hashes (Size=597), current chunk: 4, total: 53
sensor/service/connection: 2023/10/20 09:57:42.496551 connection_impl.go:704: Info: Sending 10 hashes (Size=605), current chunk: 5, total: 53
[...]
sensor/service/connection: 2023/10/20 09:57:42.507213 connection_impl.go:704: Info: Sending 10 hashes (Size=591), current chunk: 50, total: 53
sensor/service/connection: 2023/10/20 09:57:42.507590 connection_impl.go:704: Info: Sending 10 hashes (Size=603), current chunk: 51, total: 53
sensor/service/connection: 2023/10/20 09:57:42.508034 connection_impl.go:704: Info: Sending 10 hashes (Size=597), current chunk: 52, total: 53
sensor/service/connection: 2023/10/20 09:57:42.508440 connection_impl.go:704: Info: Sending 4 hashes (Size=235), current chunk: 53, total: 53
sensor/service/connection: 2023/10/20 09:57:42.508743 connection_impl.go:639: Info: Successfully sent deduper state to sensor (6cfe4d44-48d0-498a-a0a3-a0c39c832d6a)
```
```
# Sensor logs: after reconnect
common/sensor: 2023/10/20 09:57:42.564071 central_communication_impl.go:277: Info: Waiting for deduper state from Central
common/sensor: 2023/10/20 09:57:42.564264 central_communication_impl.go:311: Info: Received 10 hashes (size=603), current chunk: 1, total: 53
common/sensor: 2023/10/20 09:57:42.564378 central_communication_impl.go:311: Info: Received 10 hashes (size=587), current chunk: 2, total: 53
common/sensor: 2023/10/20 09:57:42.564496 central_communication_impl.go:311: Info: Received 10 hashes (size=607), current chunk: 3, total: 53
[...]
common/sensor: 2023/10/20 09:57:42.569448 central_communication_impl.go:311: Info: Received 10 hashes (size=591), current chunk: 50, total: 53
common/sensor: 2023/10/20 09:57:42.569540 central_communication_impl.go:311: Info: Received 10 hashes (size=603), current chunk: 51, total: 53
common/sensor: 2023/10/20 09:57:42.569639 central_communication_impl.go:311: Info: Received 10 hashes (size=597), current chunk: 52, total: 53
common/sensor: 2023/10/20 09:57:42.569715 central_communication_impl.go:311: Info: Received 4 hashes (size=235), current chunk: 53, total: 53
common/sensor: 2023/10/20 09:57:42.570104 central_communication_impl.go:188: Info: Established connection to Central.
common/sensor: 2023/10/20 09:57:42.570141 central_communication_impl.go:198: Info: Communication with central started.
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.
